File: Shell sort/ShellSort.kt
package com.exam.kotlin.codingproject

/**
 * Created by msaycon on 09,Oct,2019
 */
class ShellSort {

    companion object {

        fun sort(numbersToSort: IntArray): IntArray {
            val n = numbersToSort.size

            // Start with a big gap, then reduce the gap
            var gap = n / 2
            while (gap > 0) {
                // Do a gapped insertion sort for this gap size.
                // The first gap elements a[0..gap-1] are already
                // in gapped order keep adding one more element
                // until the entire array is gap sorted
                var i = gap
                while (i < n) {
                    // add a[i] to the elements that have been gap
                    // sorted save a[i] in temp and make a hole at
                    // position i
                    val temp = numbersToSort[i]

                    // shift earlier gap-sorted elements up until
                    // the correct location for a[i] is found
                    var j: Int
                    j = i
                    while (j >= gap && numbersToSort[j - gap] > temp) {
                        numbersToSort[j] = numbersToSort[j - gap]
                        j -= gap
                    }

                    // put temp (the original a[i]) in its correct
                    // location
                    numbersToSort[j] = temp
                    i += 1
                }
                gap /= 2
            }

            return numbersToSort
        }
    }
}

File: Shell sort/ShellSortJavascript/ShellSort.js
const shellSort = (array) => {
    var inc = array.length / 2;
    while (inc > 0) {
        for (i = inc; i < array.length; i++) {
            var j = i;
            var temp = array[i];
    
            while (j >= inc && array[j-inc] > temp) {
                array[j] = array[j-inc];
                j = j - inc;
            }
    
            array[j] = temp;
        }
    
        if (inc == 2) {
            inc = 1;
        } else {
            inc = parseInt(inc*5 / 11);
        }
    }
  return array;
}

const test = shellSort([3, 0, 2, 5, -1, 4, 1])
console.log(test);